src/notebooksyncagent.cpp

```diff
@@ -782,82 +824,193 @@ bool NotebookSyncAgent::applyRemoteChanges()
+        LOG_DEBUG("Ignore incidence already deleted locally:" << resource.href);
+        return false;
+    }
+
+    // find any existing incidence with this uid
+    QString incidenceIdentifier = QStringLiteral("%1 (%2)")
+            .arg(incidence->uid())
+            .arg(incidence->hasRecurrenceId() ? incidence->recurrenceId().toString() : QStringLiteral("original"));
+    mStorage->load(incidence->uid());
+    KCalCore::Incidence::Ptr storedIncidence;
+    switch (incidence->type()) {
+    case KCalCore::IncidenceBase::TypeEvent:
+        storedIncidence = mCalendar->event(incidence->uid(), incidence->hasRecurrenceId() ? incidence->recurrenceId() : KDateTime());
+        break;
+    case KCalCore::IncidenceBase::TypeTodo:
+        storedIncidence = mCalendar->todo(incidence->uid());
+        break;
+    case KCalCore::IncidenceBase::TypeJournal:
+        storedIncidence = mCalendar->journal(incidence->uid());
+        break;
+    case KCalCore::IncidenceBase::TypeFreeBusy:
+    case KCalCore::IncidenceBase::TypeUnknown:
+        qWarning() << "Unsupported incidence type:" << incidence->type();
+        return false;
+    }
+    if (storedIncidence) {
+        LOG_DEBUG("Updating existing event:" << incidenceIdentifier
+                                             << resource.href
+                                             << resource.etag);
+        storedIncidence->startUpdates();
+        IncidenceHandler::prepareImportedIncidence(incidence);
+        IncidenceHandler::copyIncidenceProperties(storedIncidence, incidence);
+        storedIncidence->setCustomProperty("buteo", "uri", resource.href);
+        storedIncidence->endUpdates();
+
+        // Save the modified incidence so it can be used to check whether there were
+        // local changes on the next sync.
+        mModifiedIncidenceICalData.insert(incidence->uid(), resource.iCalData);
+
+        // Save the incidence etag.
+        mUpdatedETags[incidence->uid()] = resource.etag;
+    } else {
+        LOG_DEBUG("Saving new event:" << incidenceIdentifier
+                                      << resource.href
+                                      << resource.etag);
+        KCalCore::Incidence::Ptr occurrence;
+        if (incidence->hasRecurrenceId()) {
+            // check to see if we had previously dissociated this occurrence from the recurring series.
+            occurrence = mCalendar->event(incidence->uid(), incidence->recurrenceId());
+            if (occurrence.isNull()) {
+ qWarning() << "no such recurring event occurrence?";
+                // no such dissociated occurrence existed, just create a new one.
+                // need to detach, and then copy the properties into the detached occurrence.
+                KCalCore::Incidence::Ptr recurringIncidence = mCalendar->event(incidence->uid());
+                occurrence = mCalendar->dissociateSingleOccurrence(recurringIncidence, incidence->recurrenceId(), incidence->recurrenceId().timeSpec());
```

Comment:
@pvuorela is this a correct usage of this function?  mkcal's documentation is pretty opaque.  Basically, the case we have is: we've previously stored (but perhaps not committed yet) a recurring event series, and now we have a modified occurrence (this can happen if, eg, server side we create a recurring series, and then modify the time of one occurrence, and THEN sync).

What we want is to have a single occurrence which has a recurrenceId (but the same UID) with a modified time.

Currently, when I do this, I get:

Mar 18 17:22:25 Jolla caldav-client[8997]: [W] CalDavCalendarDatabasePrivate::writeAdditions:327 - "INSERT INTO Additions (incidenceUid, notebookUid) VALUES (:incidenceUid, :notebookUid)" 
Mar 18 17:22:25 Jolla caldav-client[8997]: [W] CalDavCalendarDatabasePrivate::writeAdditions:327 - QSqlError(19, "Unable to fetch row", "UNIQUE constraint failed: Additions.incidenceUid")

because the UID is the same... but that's intentional...